File: javascript/Product.js
// javascript/Product.js
export class Product {
    constructor(name, price, quantity) {
        if (this.constructor === Product) {
            throw new Error("Abstract classes can't be instantiated.");
        }
        this.name = name;
        this.price = price;
        this.quantity = quantity;
    }

    getName() {
        return this.name;
    }

    getPrice() {
        return this.price;
    }

    getQuantity() {
        return this.quantity;
    }

    setQuantity(quantity) {
        this.quantity = quantity;
    }

    isAvailable(requestedQuantity) {
        return this.quantity >= requestedQuantity;
    }

    isExpired() {
        throw new Error("Method 'isExpired()' must be implemented.");
    }

    requiresShipping() {
        throw new Error("Method 'requiresShipping()' must be implemented.");
    }
}

File: cpp/Product.cpp
// cpp/Product.cpp
#include "Product.h"

Product::Product(std::string name, double price, int quantity)
    : name(name), price(price), quantity(quantity) {}

std::string Product::getName() const {
    return name;
}

double Product::getPrice() const {
    return price;
}

int Product::getQuantity() const {
    return quantity;
}

void Product::setQuantity(int quantity) {
    this->quantity = quantity;
}

bool Product::isAvailable(int requestedQuantity) const {
    return quantity >= requestedQuantity;
}
